tools/common-subexpr-elim.rkt

```diff
@@ -0,0 +1,67 @@
+#lang racket
+
+(require "common.rkt" "fpcore.rkt")
+
+(provide core-common-subexpr-elim)
+
+(define (core-common-subexpr-elim prog)
+  (match-define (list 'FPCore (list args ...) props ... body) prog)
+  (define cs-hash (count-common-subexpr body))
+  (define cse-body (common-subexpr-elim cs-hash body))
+  `(FPCore ,args ,@props ,cse-body))
+
+(define (combine-common-subexpr-hash h1 h2)
+  (for ([(k v) (in-hash h2)])
+    (if (hash-has-key? h1 k)
+      (hash-set! h1 k #t)
+      (hash-set! h1 k v)))
+  h1)
+
+(define (count-common-subexpr expr)
```

Comment:
Do I understand correctly that the data structure here is a map from expressions to booleans, where a value of `#f` means the expression is used once and a value of `#t` means it is used more than once? Overall the implementation here seems a little odd (lots of creating new hash tables and copying things around), but it works.